process/physics.py

```diff
@@ -379,389 +498,966 @@ def conhas(alphaj, alphap, bt, delta95, eps, kappa95, p0, rmu0):
+    - amain: np.ndarray, the atomic mass of the main ion species array
+
+    Returns:
+    - np.ndarray, the full frequency of ion collisions (Hz)
+
+    This function calculates the full frequency of ion collisions using the Coulomb logarithm of 15.
+
+    Reference:
+    - None
+    """
+    return (
+        zeff[radial_elements - 1] ** 4
+        * ni[radial_elements - 1]
+        * 322.0
+        / (
+            tempi[radial_elements - 1]
+            * np.sqrt(tempi[radial_elements - 1] * amain[radial_elements - 1])
+        )
+    )
+
+
+@nb.jit(nopython=True, cache=True)
+def ion_collisionality_sauter(
+    radial_elements: np.ndarray,
+    rmajor: float,
+    inverse_q: np.ndarray,
+    sqeps: np.ndarray,
+    tempi: np.ndarray,
+    amain: np.ndarray,
+    zeff: np.ndarray,
+    ni: np.ndarray,
+) -> float:
+    """
+    Calculate the ion collisionality to be used in the Sauter bootstrap current scaling.
+
+    Parameters:
+    - radial_elements: np.ndarray, the radial element indexes in the range 1 to nr
+    - rmajor: float, the plasma major radius (m)
+    - inverse_q: np.ndarray, inverse safety factor profile
+    - sqeps: np.ndarray, the square root of the inverse aspect ratio profile
+    - tempi: np.ndarray, the ion temperature profile (keV)
+    - amain: np.ndarray, the atomic mass of the main ion species profile
+    - zeff: np.ndarray, the effective charge of the main ion species
+    - ni: np.ndarray, the ion density profile (/m^3)
+
+    Returns:
+    - float, the ion collisionality
+
+    Reference:
+    - None
     """
     return (
         3.2e-6
-        * nui(j, zmain, ni, tempi, amain)
+        * ion_collisions_sauter(radial_elements, zeff, ni, tempi, amain)
         * rmajor
         / (
-            np.abs(mu[j - 1] + 1.0e-4)
-            * sqeps[j - 1] ** 3
-            * np.sqrt(tempi[j - 1] / amain[j - 1])
+            np.abs(inverse_q[radial_elements - 1] + 1.0e-4)
+            * sqeps[radial_elements - 1] ** 3
+            * np.sqrt(tempi[radial_elements - 1] / amain[radial_elements - 1])
         )
     )
 
 
 @nb.jit(nopython=True, cache=True)
-def dcsa(j, nr, rmajor, bt, triang, ne, ni, tempe, tempi, mu, rho, zef, sqeps):
-    """Grad(ln(ne)) coefficient in the Sauter bootstrap scaling
-    author: P J Knight, CCFE, Culham Science Centre
-    j  : input integer : radial element index in range 2 to nr
-    nr : input integer : maximum value of j
-    This function calculates the coefficient scaling grad(ln(ne))
-    in the Sauter bootstrap current scaling.
-    Code by Angioni, 29th May 2002.
-    <P>The code was supplied by Emiliano Fable, IPP Garching
-    (private communication).
-    O. Sauter, C. Angioni and Y. R. Lin-Liu,
-    Physics of Plasmas <B>6</B> (1999) 2834
-    O. Sauter, C. Angioni and Y. R. Lin-Liu, (ERRATA)
-    Physics of Plasmas <B>9</B> (2002) 5140
-
-    DCSA $\\equiv \\mathcal{L}_{31}$, Eq.14a, Sauter et al, 1999
+def calculate_l31_coefficient(
```

Comment:
Should this be an internal function (ie start the name with _)?

Reply:
Yes, changed